### PR TITLE
Conversion of prior and random trainers to estimators

### LINKED ISSUE #875: Convert Prior and Random trainers to estimators
Convert Prior and Random trainers to estimators as part of the API overhaul project, and write tests for the new estimators.

### PR DESCRIPTION
**Pull request by artidoro**
Fixes #875.

Converted prior and random trainers to estimators. 
Allowed feature column to be null for both prior and random estimator.

### CODE REVIEW

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 1]
```diff
@@ -38,38 +39,37 @@ namespace Microsoft.ML.Runtime.Learners
...
-            [Argument(ArgumentType.AtMostOnce, HelpText = "Some bool arg", ShortName = "boolarg")]
-            public bool BooleanArg = false;
         }
 
         public override PredictionKind PredictionKind => PredictionKind.BinaryClassification;
 
         private static readonly TrainerInfo _info = new TrainerInfo(normalization: false, caching: false);
         public override TrainerInfo Info => _info;
 
-        public RandomTrainer(IHostEnvironment env, Arguments args)
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >RandomTrainer [](start = 15, length = 13)

You should not remove this constructor. Once you start writing the test you will discover that it is necessary.

Add a second constructor instead. Both here and for Prior

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 2]
```diff
@@ -38,38 +39,37 @@ namespace Microsoft.ML.Runtime.Learners
     /// <summary>
     /// A trainer that trains a predictor that returns random values
     /// </summary>
-    public sealed class RandomTrainer : TrainerBase<RandomPredictor>
+
+    public sealed class RandomTrainer : TrainerEstimatorBase<BinaryPredictionTransformer<RandomPredictor>, RandomPredictor>
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >TrainerEstimatorBase [](start = 40, length = 20)

It doesn't help you to derive from this base class, and it also hurts: you now have to take label, feature and weight columns, but you don't actually need them at all.

So, don't derive from `TrainerEstimatorBase` and just implement `ITrainerEstimator`

**Comment by Zruty0** [comment]
> using Float = System.Single;

remove this line and replace all usages of `Float` with `float`

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs:5 in dda393a. [](commit_id = dda393a3ac947b98ce784090995f548d7e9de744, deletion_comment = False)

**Comment by Zruty0** [comment]
>         private readonly Random _random;

oh look, a `System.Random`. We probably want to get rid of this in favor of `TauswortheHybrid`. Check `RandomUtils.Create`

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs:99 in dda393a. [](commit_id = dda393a3ac947b98ce784090995f548d7e9de744, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 3]
```diff
@@ -38,38 +39,37 @@ namespace Microsoft.ML.Runtime.Learners
...
 
         public override PredictionKind PredictionKind => PredictionKind.BinaryClassification;
 
         private static readonly TrainerInfo _info = new TrainerInfo(normalization: false, caching: false);
         public override TrainerInfo Info => _info;
 
-        public RandomTrainer(IHostEnvironment env, Arguments args)
-            : base(env, LoadNameValue)
+        protected override SchemaShape.Column[] OutputColumns => throw new NotImplementedException();
+
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: That is obviously not sufficient. You need to list the columns that you are going to output.

Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: Same goes for the other trainer

---
In reply to: [216508453](https://github.com/dotnet/machinelearning/pull/876#discussion_r216508453) [](ancestors = 216508453)

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 4]
```diff
@@ -65,11 +63,35 @@ public RandomTrainer(IHostEnvironment env, Arguments args)
             Host.CheckValue(args, nameof(args));
         }
 
+        public BinaryPredictionTransformer<RandomPredictor> Fit(IDataView input)
+        {
+            var cachedTrain = Info.WantCaching ? new CacheDataView(Host, input, prefetch: null) : input;
+
+            RoleMappedData trainRoles = new RoleMappedData(cachedTrain);
+            var pred = Train(new TrainContext(trainRoles));
+            return new BinaryPredictionTransformer<RandomPredictor>(Host, pred, cachedTrain.Schema, trainRoles.Schema.Feature.Name);
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >trainRoles.Schema.Feature.Name [](start = 100, length = 30)

this will make features a required column, which we don't want

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 5]
```diff
@@ -65,11 +63,35 @@ public RandomTrainer(IHostEnvironment env, Arguments args)
...
+        }
+
         public override RandomPredictor Train(TrainContext context)
         {
             Host.CheckValue(context, nameof(context));
             return new RandomPredictor(Host, Host.Rand.Next());
         }
+
+        /// <summary>
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >summary [](start = 13, length = 7)

this is not a good comment. 
A good comment has to describe the purpose of the method, and not its implementation.

In this particular case, I think, it's not necessary to have a summary comment at all.

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 6]
```diff
@@ -258,6 +294,19 @@ public override PriorPredictor Train(TrainContext context)
             Float prob = prob = pos + neg > 0 ? (Float)(pos / (pos + neg)) : Float.NaN;
             return new PriorPredictor(Host, prob);
         }
+
+        protected override BinaryPredictionTransformer<PriorPredictor> MakeTransformer(PriorPredictor model, ISchema trainSchema)
+             => new BinaryPredictionTransformer<PriorPredictor>(Host, model, trainSchema, FeatureColumn.Name);
+
+        private static SchemaShape.Column MakeFeatureColumn(string featureColumn)
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >MakeFeatureColumn [](start = 42, length = 17)

for one-line methods use `=>` syntax.

**Comment by Zruty0** [comment]
Still waiting for tests.

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs [hunk 7]
```diff
@@ -91,12 +91,36 @@ public IList<TestDataset> GetDatasetsForClassificationWeightingPredictorsTest()
         [TestCategory("Binary")]
         public void BinaryClassifierPerceptronTest()
         {
-            var binaryPredictors = new[] { TestLearners.perceptron };
+            var binaryPredictors = new[] { TestLearners.perceptron};
```
Comment (Zruty0) [inline test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: >on}; [](start = 64, length = 4)

undo this change

File: test/BaselineOutput/SingleDebug/RandomPredictor/BinaryRandom-TrainTest-breast-cancer-test-out.txt [hunk 8]
```diff
@@ -0,0 +1,18 @@
+maml.exe Test dout=%Output% data=%Data% in=%Output% seed=1
+Physical memory usage(MB): %Number%
+Virtual memory usage(MB): %Number%
+%DateTime%	 Time elapsed(s): %Number%
+
+***** Unexpected failure. Please refer to https://aka.ms/MLNetIssue to file an issue with details *****
+***** Error log has been saved to '%Temp%\%ErrorLog%', please refer to https://aka.ms/MLNetIssue to file an issue with details *****
+===== Begin detailed dump =====
+(1) Unexpected exception: Object reference not set to an instance of an object., 'System.NullReferenceException'
```
Comment (Zruty0) [inline test/BaselineOutput/SingleDebug/RandomPredictor/BinaryRandom-TrainTest-breast-cancer-test-out.txt]: >Unexpected exception [](start = 4, length = 20)

this is not a correct behavior

File: test/BaselineOutput/SingleDebug/PriorPredictor/BinaryPrior-TrainTest-breast-cancer-test-out.txt [hunk 9]
```diff
@@ -0,0 +1,19 @@
+maml.exe Test dout=%Output% data=%Data% in=%Output% seed=1
+Physical memory usage(MB): %Number%
+Virtual memory usage(MB): %Number%
+%DateTime%	 Time elapsed(s): %Number%
+
+***** Unexpected failure. Please refer to https://aka.ms/MLNetIssue to file an issue with details *****
+***** Error log has been saved to '%Temp%\%ErrorLog%', please refer to https://aka.ms/MLNetIssue to file an issue with details *****
+===== Begin detailed dump =====
+(1) Unexpected exception: Could not find file '%Output% 'System.IO.FileNotFoundException'
```
Comment (Zruty0) [inline test/BaselineOutput/SingleDebug/PriorPredictor/BinaryPrior-TrainTest-breast-cancer-test-out.txt]: > Unexpected exception [](start = 3, length = 21)

make sure your code doesn't throw exceptions

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs (hunk 4 above)
Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: Let's follow through and formally say that feature column is also optional in `PredictionTransformerBase`: right now it is not

---
In reply to: [216845418](https://github.com/dotnet/machinelearning/pull/876#discussion_r216845418) [](ancestors = 216845418)

File: src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs [hunk 10]
```diff
@@ -36,17 +36,18 @@ public abstract class PredictionTransformerBase<TModel> : IPredictionTransformer
 
         public TModel Model { get; }
 
-        public PredictionTransformerBase(IHost host, TModel model, ISchema trainSchema, string featureColumn)
+        public PredictionTransformerBase(IHost host, TModel model, ISchema trainSchema, string featureColumn = null)
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs]: > = null [](start = 108, length = 7)

nope, that should still be required param. Just 'null' should now be acceptable.

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 11]
```diff
@@ -13,6 +11,8 @@
 using Microsoft.ML.Runtime.Model;
 using Microsoft.ML.Runtime.Training;
 using Microsoft.ML.Runtime.Internal.Internallearn;
+using Microsoft.ML.Core.Data;
+using System.Linq;
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >System [](start = 6, length = 6)

sort namespaces

**Comment by Zruty0** [comment]
Now waiting for test containing `TestEstimatorCore`

---
In reply to: [420450005](https://github.com/dotnet/machinelearning/pull/876#issuecomment-420450005) [](ancestors = 420450005)

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs (hunk 4 above)
Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: don't make it an optional parameter

---
In reply to: [217102915](https://github.com/dotnet/machinelearning/pull/876#discussion_r217102915) [](ancestors = 217102915,216845418)

File: test/Microsoft.ML.Predictor.Tests/TestEstimators.cs [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.EntryPoints;
+using Microsoft.ML.Runtime.Learners;
+using Microsoft.ML.Runtime.RunTests;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Tests
+{
+    public class SimpleEstimatorTests : TestDataPipeBase
```
Comment (Zruty0) [inline test/Microsoft.ML.Predictor.Tests/TestEstimators.cs]: >SimpleEstimatorTests [](start = 17, length = 20)

The filename is overly broad. Let's make it something like `TestTrainerEstimators`

File: test/Microsoft.ML.Predictor.Tests/TestTrainerEstimators.cs [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.EntryPoints;
+using Microsoft.ML.Runtime.Learners;
+using Microsoft.ML.Runtime.RunTests;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Tests
+{
+    public class SimpleEstimatorTests : TestDataPipeBase
```
Comment (sfilipi) [inline test/Microsoft.ML.Predictor.Tests/TestTrainerEstimators.cs]: >s SimpleEstimat [](start = 15, length = 15)

move them to the TraienrEstimators folder, so we have them all in one place.

File: test/Microsoft.ML.Predictor.Tests/TestTrainerEstimators.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        [Fact]
+        public void TestEstimatorRandom()
+        {
+            var dataView = GetBreastCancerDataviewWithTextColumns();
+            var pipe = new RandomTrainer(Env, new RandomTrainer.Arguments());
+
+            // Test only that the schema propagation works.
+            // REVIEW: the save/load is not preserving the full state of the random predictor. This is unfortunate, but we don't care too much at this point.
```
Comment (sfilipi) [inline test/Microsoft.ML.Predictor.Tests/TestTrainerEstimators.cs]: >REVIEW: the save/l [](start = 15, length = 18)

maybe log an issue after merging

Comment (artidoro) [reply test/Microsoft.ML.Predictor.Tests/TestTrainerEstimators.cs]: Ok will do that!

---
In reply to: [218139123](https://github.com/dotnet/machinelearning/pull/876#discussion_r218139123) [](ancestors = 218139123)

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 15]
```diff
@@ -38,20 +38,16 @@ namespace Microsoft.ML.Runtime.Learners
     /// <summary>
     /// A trainer that trains a predictor that returns random values
     /// </summary>
-    public sealed class RandomTrainer : TrainerBase<RandomPredictor>
+
+    public sealed class RandomTrainer : TrainerBase<RandomPredictor>,
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >TrainerBase<RandomPredictor> [](start = 40, length = 28)

I think this can go away too. It will eventually, to my understanding. @Zruty0 can confirm.

**Comment by sfilipi** [comment]
>         public RandomTrainer(IHostEnvironment env, Arguments args)

private. 

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs:58 in 3dcb705. [](commit_id = 3dcb7057e952e84ec90a0969814cab62b030af99, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 16]
```diff
@@ -38,20 +38,16 @@ namespace Microsoft.ML.Runtime.Learners
...
+        ITrainerEstimator<BinaryPredictionTransformer<RandomPredictor>, RandomPredictor>
     {
         internal const string LoadNameValue = "RandomPredictor";
         internal const string UserNameValue = "Random Predictor";
         internal const string Summary = "A toy predictor that returns a random value.";
 
-        public class Arguments
+        public sealed class Arguments
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >public [](start = 8, length = 6)

as you go through those, try adding XML comments to everything public, as they display on the docs site: docs.microsoft.com

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 17]
```diff
@@ -211,12 +236,16 @@ public sealed class Arguments
         public override TrainerInfo Info => _info;
 
         public PriorTrainer(IHostEnvironment env, Arguments args)
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >public [](start = 8, length = 6)

private

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 18]
```diff
@@ -211,12 +236,16 @@ public sealed class Arguments
...
-            : base(env, LoadNameValue)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(LoadNameValue), MakeFeatureColumn(DefaultColumnNames.Features), MakeLabelColumn(DefaultColumnNames.Label), null)
+        {
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >{ [](start = 8, length = 1)

keep the args check.

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs (hunk 17 above)
Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: Why should I make the constructor private?

---
In reply to: [218154524](https://github.com/dotnet/machinelearning/pull/876#discussion_r218154524) [](ancestors = 218154524)

**Comment by artidoro** [comment]
>         public RandomTrainer(IHostEnvironment env, Arguments args)

Same question as bello: Why should this be private? Isn't this the only constructor of RandomTrainer?

---
In reply to: [422096175](https://github.com/dotnet/machinelearning/pull/876#issuecomment-422096175) [](ancestors = 422096175)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs:58 in 3dcb705. [](commit_id = 3dcb7057e952e84ec90a0969814cab62b030af99, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs (hunk 15 above)
Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: Yes, this is an exception :)

---
In reply to: [218156771](https://github.com/dotnet/machinelearning/pull/876#discussion_r218156771) [](ancestors = 218156771,218152453)

**Comment by Zruty0** [comment]
>         public RandomTrainer(IHostEnvironment env, Arguments args)

you should actually have a parameterless constructor, and this one needs to be present just for the component catalog, and it can be made private now.

---
In reply to: [422106631](https://github.com/dotnet/machinelearning/pull/876#issuecomment-422106631) [](ancestors = 422106631,422096175)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs:58 in 3dcb705. [](commit_id = 3dcb7057e952e84ec90a0969814cab62b030af99, deletion_comment = False)

File: test/Microsoft.ML.Predictor.Tests/TestTrainerEstimators.cs [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+            TestEstimatorCore(pipe, new EmptyDataView(Env, dataView.Schema));
+            Done();
+        }
+
+        [Fact]
+        public void TestEstimatorPrior()
+        {
+            using (var env = new TlcEnvironment())
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Predictor.Tests/TestTrainerEstimators.cs]: >env [](start = 23, length = 3)

Use baseclass Env, no need to create another one.

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 20]
```diff
@@ -211,12 +236,16 @@ public sealed class Arguments
...
+        }
+
+        public PriorTrainer(IHost host, SchemaShape.Column feature, SchemaShape.Column label, SchemaShape.Column weight)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >public [](start = 8, length = 6)

Can it be private?
I understand what you use TrainerEstimatorBase which requires feature column, but for this class it's pointless, and exposing it to user would be potentially confusing.

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 21]
```diff
@@ -234,16 +263,16 @@ public override PriorPredictor Train(TrainContext context)
             using (var cursor = data.Data.GetRowCursor(c => c == col || c == colWeight))
             {
                 var getLab = cursor.GetLabelFloatGetter(data);
-                var getWeight = colWeight >= 0 ? cursor.GetGetter<Float>(colWeight) : null;
-                Float lab = default(Float);
-                Float weight = 1;
+                var getWeight = colWeight >= 0 ? cursor.GetGetter<float>(colWeight) : null;
+                float lab = default(float);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >float [](start = 36, length = 5)

you can omit it.

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 22]
```diff
@@ -210,13 +243,25 @@ public sealed class Arguments
...
             Host.CheckValue(args, nameof(args));
         }
 
-        public override PriorPredictor Train(TrainContext context)
+        /// <summary>
+        /// Initializes PriorTrainer object.
+        /// </summary>
+        /// <param name="host"></param>
+        /// <param name="feature"></param>
+        /// <param name="label"></param>
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: either omit, or put something into them

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 23]
```diff
@@ -183,20 +216,20 @@ private Float PredictCore()
...
 
-        private void MapDist(ref VBuffer<Float> src, ref Float score, ref Float prob)
+        private void MapDist(ref VBuffer<float> src, ref float score, ref float prob)
         {
             score = PredictCore();
             prob = (score + 1) / 2;
         }
     }
 
     // Learns the prior distribution for 0/1 class labels and just outputs that.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >// [](start = 4, length = 2)

Make it ///<Summary>

File: src/Microsoft.ML.Core/Data/RoleMappedSchema.cs [hunk 24]
```diff
@@ -534,7 +534,7 @@ public RoleMappedData(IDataView data, IEnumerable<KeyValuePair<RoleMappedSchema.
         /// <param name="opt">Whether to consider the column names specified "optional" or not. If <c>false</c> then any non-empty
         /// values for the column names that does not appear in <paramref name="data"/>'s schema will result in an exception being thrown,
         /// but if <c>true</c> such values will be ignored</param>
-        public RoleMappedData(IDataView data, string label, string feature,
+        public RoleMappedData(IDataView data, string label, string feature = null,
```
Comment (Zruty0) [inline src/Microsoft.ML.Core/Data/RoleMappedSchema.cs]: > = null [](start = 74, length = 7)

undo this

Comment (artidoro) [reply src/Microsoft.ML.Core/Data/RoleMappedSchema.cs]: I will handle this in my code then! Sorry about that

---
In reply to: [218244002](https://github.com/dotnet/machinelearning/pull/876#discussion_r218244002) [](ancestors = 218244002)

File: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs [hunk 25]
```diff
@@ -216,6 +254,28 @@ public PriorTrainer(IHostEnvironment env, Arguments args)
             Host.CheckValue(args, nameof(args));
         }
 
+        /// <summary>
+        /// Initializes PriorTrainer object.
+        /// </summary>
+        public PriorTrainer(IHost host, SchemaShape.Column label, SchemaShape.Column weight)
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: >SchemaShape.Column label [](start = 40, length = 24)

The constructor signature should be 
`PriorTrainer(IHostEnvironment env, string labelColumn, string weightColumn = null)`

Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs]: Ok, I made those Strings instead of SchemShape.Column

---
In reply to: [218244513](https://github.com/dotnet/machinelearning/pull/876#discussion_r218244513) [](ancestors = 218244513)

File: test/Microsoft.ML.Tests/TrainerEstimators/PriorRandomTests.cs [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+                    }
+                },
+
+                InputFile = inputFile
+            }).Data;
+        }
+
+        private static SchemaShape.Column MakeLabelColumn(string labelColumn)
+            => new SchemaShape.Column(labelColumn, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false);
+
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/TrainerEstimators/PriorRandomTests.cs]: don't think you're using this

File: test/Microsoft.ML.Tests/TrainerEstimators/PriorRandomTests.cs [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        [Fact]
+        public void TestEstimatorRandom()
+        {
+            var dataView = GetBreastCancerDataviewWithTextColumns();
+            var pipe = new RandomTrainer(Env);
+
+            // Test only that the schema propagation works.
+            // REVIEW: the save/load is not preserving the full state of the random predictor. This is unfortunate, but we don't care too much at this point.
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/TrainerEstimators/PriorRandomTests.cs]: Is it still the case?
If yes, can you write down what "full" state are we missing?

Comment (artidoro) [reply test/Microsoft.ML.Tests/TrainerEstimators/PriorRandomTests.cs]: This comment was written by Pete. I don't know exactly what "full state" refers to. I am going to search more in depth to understand what is causing the problem and file a separate issue if it is needed. The behavior that we observed is that the data produced by the model that was saved + loaded back is different from the data produced by the original model.

---
In reply to: [218520614](https://github.com/dotnet/machinelearning/pull/876#discussion_r218520614) [](ancestors = 218520614)
